### PR TITLE
Recover from FLAC decoding errors

### PR DESCRIPTION
**Pull request by uklotzde**
Retested corrupt files and fixed some hiccups.

### CODE REVIEW

**Comment by ronso0** [comment]
By chance, is this also fixing https://bugs.launchpad.net/mixxx/+bug/1846409 ?
Did @daschuer forward my affected rack to you?
If not I can do so. (Don't have the ressources to built this branch and test it myself right now)

**Comment by uklotzde** [comment]
> By chance, is this also fixing https://bugs.launchpad.net/mixxx/+bug/1846409 ?
> Did @daschuer forward my affected track to you?
> If not I can do so. (Don't have the ressources to built this branch and test it myself right now)

No, FLAC decoding worked perfectly. This PR only fixes issues with really bad and corrupt files.

**Comment by ronso0** [comment]
My mistake, encoding/decoding :)

**Comment by uklotzde** [comment]
@ronso0 Nevertheless please forward the file also to me (Zulip link or e-mail), just in case.

File: src/sources/soundsourceflac.cpp [hunk 1]
```diff
@@ -193,8 +199,12 @@ ReadableSampleFrames SoundSourceFLAC::readSampleFramesClamped(
                 && (precedingFrames != readSampleFramesClamped(
                         WritableSampleFrames(precedingFrames)).frameIndexRange())) {
             kLogger.warning()
-                    << "Failed to skip preceding frames"
+                    << "Resetting decoder after failure to skip preceding frames"
                     << precedingFrames;
+            if (!FLAC__stream_decoder_reset(m_decoder)) {
+                kLogger.critical()
```
Comment (daschuer) [inline src/sources/soundsourceflac.cpp]: This crashes Mixxx, right? Doe we have alternatives?

Comment (uklotzde) [reply src/sources/soundsourceflac.cpp]: Not by default: "It exits if the environment variable QT_FATAL_CRITICALS is not empty."

Comment (uklotzde) [reply src/sources/soundsourceflac.cpp]: If even resetting the decoder fails then we are in big trouble. Like a destructor this operation should never fail, but the API allows failure and we need to handle this case.

Comment (uklotzde) [reply src/sources/soundsourceflac.cpp]: Playback was stuttering and Mixxx became almost unusable when not resetting the decoder. This is already our very last chance. If even this option fails everything is lost.

This is still only a DJ application, not the control unit of some deadly weapon. Let's keep things in perspective.

**Comment by daschuer** [comment]
LGTM. Thank you.

**Pull request merged**